compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -222,52 +208,28 @@ impl<'s> AllocDecodingSession<'s> {
-                State::Done(alloc_id) => {
-                    return alloc_id;
-                }
-                ref mut entry @ State::Empty => {
-                    // We are allowed to decode.
-                    match alloc_kind {
-                        AllocDiscriminant::Alloc => {
-                            // If this is an allocation, we need to reserve an
-                            // `AllocId` so we can decode cyclic graphs.
-                            let alloc_id = decoder.interner().reserve_alloc_id();
-                            *entry = State::InProgress(smallvec![self.session_id], alloc_id);
-                            Some(alloc_id)
-                        }
-                        AllocDiscriminant::Fn
-                        | AllocDiscriminant::Static
-                        | AllocDiscriminant::VTable => {
-                            // Fns and statics cannot be cyclic, and their `AllocId`
-                            // is determined later by interning.
-                            *entry = State::InProgressNonAlloc(smallvec![self.session_id]);
-                            None
-                        }
+        let mut entry = self.state.decoding_state[idx].lock();
+        let alloc_id = match *entry {
+            State::Done(alloc_id) => {
+                return alloc_id;
+            }
+            State::Empty => {
+                // We are allowed to decode.
+                match alloc_kind {
+                    AllocDiscriminant::Alloc => {
+                        // If this is an allocation, we need to reserve an
+                        // `AllocId` so we can decode cyclic graphs.
+                        let alloc_id = decoder.interner().reserve_alloc_id();
+                        Some(alloc_id)
                     }
-                }
-                State::InProgressNonAlloc(ref mut sessions) => {
-                    if sessions.contains(&self.session_id) {
-                        bug!("this should be unreachable");
-                    } else {
-                        // Start decoding concurrently.
-                        sessions.push(self.session_id);
+                    AllocDiscriminant::Fn
+                    | AllocDiscriminant::Static
+                    | AllocDiscriminant::VTable => {
+                        // Fns and statics cannot be cyclic, and their `AllocId`
+                        // is determined later by interning.
                         None
                     }
                 }
-                State::InProgress(ref mut sessions, alloc_id) => {
-                    if sessions.contains(&self.session_id) {
-                        // Don't recurse.
-                        return alloc_id;
-                    } else {
-                        // Start decoding concurrently.
-                        sessions.push(self.session_id);
-                        Some(alloc_id)
-                    }
-                }
             }
```

Comment:
This is not necessary anymore. Just reserve the id where the option is unwrapped below

Reply:
Oh! I see now